### PR TITLE
[Release scripting] Use native boto3 session interface for AWS authentication

### PR DESCRIPTION
**Pull request by mellenburg**
## High-level description
Before this change, the release script had to use explicit environment variables for access key ID and secret access key in order to authenticate with AWS but this is less than ideal and instead we should just use the AWS built-in interface. For example, with this change, using AWS_SESSION_TOKEN is an option

## Corresponding DC/OS tickets

https://jira.mesosphere.com/browse/DCOS_OSS-2064

## Checklist for all PRs

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

### CODE REVIEW

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

**Comment by mellenburg** [comment]
@lingmann I think this is ready for review now, so PTAL

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

File: dcos_installer/test_backend.py [hunk 1]
```diff
@@ -288,7 +265,7 @@ def test_do_configure(tmpdir, monkeypatch):
 """
 
 
-def test_do_aws_configure(tmpdir, monkeypatch):
+def test_do_aws_configure(tmpdir, monkeypatch, release_config_aws):
```
Comment (orsenthil) [inline dcos_installer/test_backend.py]: argument order. There could be an intuitive preference for `monkeypatch` arg to be the last one.

File: dcos_installer/test_backend.py [hunk 2]
```diff
@@ -297,54 +274,50 @@ def test_do_aws_configure(tmpdir, monkeypatch):
         assert backend.do_aws_cf_configure() == 0
 
 
-valid_storage_config = """---
+@pytest.fixture
+def valid_storage_config(release_config_aws):
+    """ Uses the settings from dcos-release.config.yaml ['testing'] to create a
+    new upload and then deletes it when the test is over
+    """
+    s3_bucket_name = release_config_aws['bucket']
```
Comment (orsenthil) [inline dcos_installer/test_backend.py]: `release_config_aws.get('bucket', '')` ?  Or do you want this to fail with `KeyError` ?

File: dcos_installer/test_backend.py [hunk 3]
```diff
@@ -297,54 +274,50 @@ def test_do_aws_configure(tmpdir, monkeypatch):
...
+    bucket_path = str(uuid.uuid4())
```
Comment (orsenthil) [inline dcos_installer/test_backend.py]: `uuid.uuid4().encode('ascii')` - please.

File: release/storage/aws.py [hunk 4]
```diff
@@ -6,32 +6,38 @@
...
-                aws_access_key_id=access_key_id,
-                aws_secret_access_key=secret_access_key,
-                region_name=region_name)
-        else:
-            raise ValueError("boto_profile or explicit AWS credentials (access_key_id, "
-                             "secret_access_key, region_name) must be provided")
+def get_aws_session(access_key_id, secret_access_key, region_name=None):
+    """ This method will replace access_key_id and secret_access_key
+    with None if one is set to '' This allows falling back to the AWS internal
+    logic so that AWS_SESSION_TOKEN or something else can be used
```
Comment (orsenthil) [inline release/storage/aws.py]: What does `something else` refer to here?
`some else like IAM role` ?

Comment (orsenthil) [reply release/storage/aws.py]: or link the `http://boto3.readthedocs.io/en/latest/guide/configuration.html#configuring-credentials` again.

File: tox.ini [hunk 5]
```diff
@@ -41,17 +41,15 @@ passenv =
...
+  AWS_ACCESS_KEY_ID
+  AWS_SECRET_ACCESS_KEY
+  AWS_SESSION_TOKEN
+  AWS_PROFILE
+  AWS_CONFIG_FILE
   AWS_DEFAULT_REGION
 deps =
   dnspython
   pytest
-  pytest-catchlog
```
Comment (orsenthil) [inline tox.ini]: Why is this removed?

**Comment by orsenthil** [comment]
@mellenburg  - In the testing done session, could you please share the output behavior of  the `release` script after this change is made. 
```
release create some-version mellenburg/dcos/some-version
```

and verify that the artifacts are uploaded correctly. 

* This way we can ensure that the release behavior hasn't regressed.

File: conftest.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+@pytest.fixture
+def release_config_testing(release_config):
+    if 'testing' not in release_config:
+        pytest.skip("Skipped because there is no `testing` configuration in dcos-release.config.yaml")
+    return release_config['testing']
+
+
+@pytest.fixture
+def release_config_aws(release_config_testing, monkeypatch):
```
Comment (orsenthil) [inline conftest.py]: `monkeypatch` arg kind of indicates that only this method is exercised/ tested in code. 
Perhaps others need to be exercised and tested too.

File: dcos_installer/test_backend.py [hunk 7]
```diff
@@ -297,54 +274,50 @@ def test_do_aws_configure(tmpdir, monkeypatch):
...
 
 
-def test_do_aws_cf_configure_valid_storage_config(config_aws, tmpdir, monkeypatch):
-    bucket = str(uuid.uuid4())
-    config_str = valid_storage_config.format(
-        key_id=config_aws["access_key_id"],
-        bucket=bucket,
-        access_key=config_aws["secret_access_key"])
-    assert aws_cf_configure(bucket, config_str, config_aws, tmpdir, monkeypatch) == 0
+def test_do_aws_cf_configure_valid_storage_config(release_config_aws, tmpdir, monkeypatch, valid_storage_config):
```
Comment (orsenthil) [inline dcos_installer/test_backend.py]: same here with w.r.t order of args. We could consider keeping the patcher arg to the last one.

File: dcos_installer/test_backend.py [hunk 8]
```diff
@@ -297,54 +274,50 @@ def test_do_aws_configure(tmpdir, monkeypatch):
...
 
-def test_override_aws_template_storage_region_name(config_aws, tmpdir, monkeypatch):
-    bucket = str(uuid.uuid4())
-    config_str = valid_storage_config.format(
-        key_id=config_aws["access_key_id"],
-        bucket=bucket,
-        access_key=config_aws["secret_access_key"])
-    config_str += '\naws_template_storage_region_name: {}'.format(config_aws['region_name'])
-    assert aws_cf_configure(bucket, config_str, config_aws, tmpdir, monkeypatch) == 0
+def test_override_aws_template_storage_region_name(release_config_aws, tmpdir, monkeypatch, valid_storage_config):
```
Comment (orsenthil) [inline dcos_installer/test_backend.py]: ditto.

File: release/storage/aws.py [hunk 9]
```diff
@@ -6,32 +6,38 @@
...
+        region_name=region_name)
 
 
 class S3StorageProvider(AbstractStorageProvider):
     name = 'aws'
 
-    def __init__(self, bucket, object_prefix, download_url, boto3_profile=None, region_name=None,
-                 access_key_id=None, secret_access_key=None):
+    def __init__(self, bucket, object_prefix, download_url,
+                 access_key_id=None, secret_access_key=None, region_name=None):
```
Comment (orsenthil) [inline release/storage/aws.py]: What if the `default-region` is None. AFAICS, it takes it from the environment / config file of the user. The code inside boto3 does not have any default region.

So, perhaps it might be useful to state this / having  default region in `our code` set to us-west2?

File: dcos_installer/test_backend.py (hunk 2 above)
Comment (mellenburg) [reply dcos_installer/test_backend.py]: You most definitely want this to fail with a key error. Running this without a pre-configured bucket for writing to not a valid test case (as the advanced aws template doesnt itself create a new bucket, but rather just a bucket path)

File: tox.ini [hunk 10]
```diff
@@ -41,17 +41,15 @@ passenv =
...
-  AWS_TESTING_SECRET_ACCESS_KEY
-  AWS_PROD_ACCESS_KEY_ID
-  AWS_PROD_SECRET_ACCESS_KEY
-  AWS_DEV_ACCESS_KEY_ID
-  AWS_DEV_SECRET_ACCESS_KEY
+  AWS_ACCESS_KEY_ID
+  AWS_SECRET_ACCESS_KEY
+  AWS_SESSION_TOKEN
+  AWS_PROFILE
+  AWS_CONFIG_FILE
```
Comment (orsenthil) [inline tox.ini]: It will be great, if we have a comment with references in this file. I think, `.ini` files allow it.
If someone starts the read from the `.ini` file, they are going to miss the understanding of the behavior and fallbacks that is used in the code.

File: dcos_installer/test_backend.py (hunk 1 above)
Comment (mellenburg) [reply dcos_installer/test_backend.py]: by that same logic, wouldn't `tmpdir` also have to be behind `release_config_aws`? Seems inconsequential unless we have a rule that says "all built-in pytest fixtures should be listed after custom fixtures"

File: tox.ini (hunk 5 above)
Comment (mellenburg) [reply tox.ini]: because when you run pytest/tox you see this:
`pytest-catchlog plugin has been merged into the core, please remove it from your requirements`
https://teamcity.mesosphere.io/viewLog.html?buildId=931198&buildTypeId=DcOs_Open_Build_Tox&tab=buildLog&_focus=1523#_state=268

**Comment by mellenburg** [comment]
@orsenthil I do not understand your request that I do `release create foo bar` our of band and verify the artifacts... This is literally exactly what the CI does...
- teamcity/dcos/build/dcos runs build_teamcity like this https://github.com/dcos/dcos/blob/master/build_teamcity#L80
- teamcity/dcos/test/aws/onprem/static pulls the generated dcos_generate_config.sh from the upload bucket
- teamcity/dcos/test/aws/cloudformation/simple pulls the simple templates and launches them from the upload bucket
- teamcity/dcos/test/azure/arm pulls the ACS templates from the upload bucket and runs them

I fail to see what I should be doing that isn't already covered in CI

File: release/storage/aws.py (hunk 9 above)
Comment (mellenburg) [reply release/storage/aws.py]: I oppose this. We would *much* rather see an error to set the default region than to quietly set it somewhere buried in release/storage/aws.py

File: conftest.py (hunk 6 above)
Comment (mellenburg) [reply conftest.py]: monkeypatch was a holdover from a previous revision and should be removed now. good eye

**Comment by orsenthil** [comment]
@mellenburg  - I might have "release promote" (followed by release create) in my mind. We certainly don't do `release promote` in CI.

File: dcos_installer/test_backend.py (hunk 3 above)
Comment (mellenburg) [reply dcos_installer/test_backend.py]: I'm confused... UUID object does not natively have an 'encode' method, so are you asking me to do `str(uuid.uuid4()).encode('ascii')`? That serves no purpose because that will have to be decoded as a python string when tit is used later on... This seems like a very unnecessary request.

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

**Comment by mellenburg** [comment]
@orsenthil I verified release promotion with the following:
```
release promote testing/pull/2294 testing/foobar/release
```

File: dcos_installer/test_backend.py (hunk 3 above)
Comment (orsenthil) [reply dcos_installer/test_backend.py]: My mistake. I mistakenly assumed `uuid4()` returned a byte-string.

File: dcos_installer/test_backend.py [hunk 11]
```diff
@@ -297,54 +274,50 @@ def test_do_aws_configure(tmpdir, monkeypatch):
...
+""".format(
+        bucket=release_config_aws['bucket'],
+        bucket_path=bucket_path)
+    session = boto3.session.Session()
+    s3 = session.resource('s3')
+    s3_bucket = s3.Bucket(s3_bucket_name)
+    for o in s3_bucket.objects.filter(Prefix=bucket_path):
+        o.delete()
+    # objects = [{'Key': o.key} for o in s3_bucket.objects.filter(Prefix=bucket_path)]
+    # s3_bucket.delete_objects(Delete={'Objects': objects})
```
[the remark this answers is not in the document]
Comment (mellenburg) [reply dcos_installer/test_backend.py]: Oops, thanks. Good eye!

**Comment by mellenburg** [comment]
@mesosphere-mergebot bump-ee

File: release/storage/aws.py (hunk 9 above)
Comment (deric) [reply release/storage/aws.py]: I don't find it very useful to see an AWS region error when trying to build a local release, that I'm not deploying to AWS. Is there an option to disable boto3 completely?
